### ISSUE TITLE
Optimize HTML-import

### ISSUE THREAD

**Issue by Morik** [WoltLab/com.woltlab.wcf.exporter#25]
Currently HTML-posts and other messages are imported "as they are" to prevent further processing.
But this has a few side effects like not being able to rebuild posts in order to re-detect bbcodes after installing a new package etc.
This may be seams like not a big deal, but it's frustrating as many times you see just blank bbcodes until you save them and if you want to edit a HTML post it ends, most of the time, in absolute chaos.
So i suggest to wrap such posts inside the new html-bbcode and enable reprocessing posts etc throw the rebuilt workers.

**Comment by dtdesign** [comment]
> So i suggest to wrap such posts inside the new html-bbcode and enable reprocessing posts etc throw the rebuilt workers.

This won't solve your issue because content inside the `[html]`-bbcode isn't processed by the parser at all.

**Comment by Morik** [comment]
Why not ?
It gets converted to a pre tag and it's content will then be ignored by HTMLPurifyer, or did i miss something ?
I thought this was the reason to implement the new HTML bbcode in the first place.
Maybe its nessecary to run such posts throu htmlspecialchars befor wrapping it inside the bbcode (didn't checked this) but i don't see a big problem at the moment for such a change.

**Comment by dtdesign** [comment]
The contents of a `<pre>` is treated as plain text, bbcodes inside are not recognized at all.

**Comment by Morik** [comment]
OK, maybe i wasn't that clear about what i want to achieve.
I want that imports from versions below WSF3 are handled differently depending on their html tag.
If the post was written as bbcode the content should be importet like currently and if its html it should be wrapped.
This way you can edit HTML posts without to worry if the content gets lost after saving it.
And with this change you could enable reparsing of posts with the html tag by the worker without loosing pure html posts.
I thought this was the reason to limit the worker to non html posts.

**Comment by dtdesign** [comment]
We can wrap them into the HTML bbcode, which makes it safer to edit them (e. g. imported via RSS feed).

However, reparsing would still do not change **anything**, no matter if these HTML posts are wrapped in the HTML bbcode or not.

**Comment by Morik** [comment]
Well, let me give you an example:
An Admin is using an bbcode like `[info]` and imports the post from phpbb to wsf (or even from wbb4 etc, doesn't really matter).
Following the guide he is rebuilding all data sets and has now plain `[info]some text[/info]` inside his posts.
He goes then to the sore and loads the wsf version of this bbcode in order to fix his posts.
But after installing the plugin nothing happens and he still sees just the plain text until he edits the post.
The problem is, that unknown bbcodes aren't transformed to metacodes.
Rebuilding the posts won't also help as all the posts are marked as html posts and therefore not beeing reparsed.
This is frustrating and it was neccessary to modify the worker for some users (adding the ability to reparse posts) in order to fix their problems.

**Comment by dtdesign** [comment]
I understood your intentions, but again, using the HTML bbcode does not change the behavior when rebuilding posts. The `[info]` bbcode won't be recognized, because either the posts has already been processed (`enableHtml = 1`) or it sits inside the HTML bbcode. You do realize that bbcodes are not evaluated inside the HTML bbcode, right?

**Comment by jens1o** [comment]
> You do realize that bbcodes are not evaluated inside the HTML bbcode, right?

But I would like they do...

**Comment by Morik** [comment]
> because either the posts has already been processed (`enableHtml = 1`)
I want to change this to, by protecting pure html posts this shouldn't be a problem and would this way solve the problem.

**Comment by dtdesign** [comment]
@Morik I'm sorry, but you've taken the wrong turns. The problem outlined by you cannot be solved by wrapping the content into the HTML bbcode.

To be absolute clear: Not a single change that was suggested in this issue is going to have an impact of the bbcode processing.

**Comment by Morik** [comment]
Lets try it step by step, maybe this shows you my real intention.

The first thing i would do is wrapping real html posts (not wsf5 html posts which are already processed) while import into the html bbcode in order to protect the content.
This shouldn't make a difference in the result as these content is later shown as it is as if it wasn't wrapped in the first place.

If all imported html posts (also conversations etc) are wrapped this means those posts can be processed (which means running it thru the `HtmlInputProcessor`) by the rebuild worker without loosing content.
So if there is no data loss with reprocessing this could be changed in order to allow redetecting of bbcodes unknown to the system while the first processing.
The last point currently fails because only embedded content is reprocessed and not the whole post.

**Comment by dtdesign** [comment]
> The first thing i would do is wrapping real html posts (not wsf5 html posts which are already processed) while import into the html bbcode in order to protect the content.

We can do that.

> If all imported html posts (also conversations etc) are wrapped this means those posts can be processed (which means running it thru the `HtmlInputProcessor`) by the rebuild worker without loosing content.

This is correct.

> So if there is no data loss with reprocessing this could be changed in order to allow redetecting of bbcodes unknown to the system while the first processing.

… and that doesn't work. Because of your suggested change, the original post is now wrapped in the HTML bbcode, which means that __ALL__ bbcodes contained inside it will not be detected. The bbcode parser will see pretty much this: `[html]…[/html]`, where `…` represents the wrapped content that (by definition of code-type bbcodes) contains absolutely no bbcodes and never will.

The bottom line is: If the post is marked as HTML in the source, any contained bbcodes will never be recognized without manual editing. Your suggested change has zero impact on this behavior!

**Comment by Morik** [comment]
I don't want to detect bbcodes inside the html tag.
This step was only necessary to be able to reparse normal bbcode posts a second time while protecting real html posts.

**Comment by dtdesign** [comment]
> This step was only necessary to be able to reparse normal bbcode posts a second time while protecting real html posts.

Well then, this is still not possible with your suggested change, because it simply does not work this way. Rebuilding posts will parse the contained data, but will not re-parse the message itself. This is different in a way that it will only deconstruct the message while searching for certain elements, but no actual conversion is performed, this includes the evaluation of bbcode tags.

**Comment by Morik** [comment]
> Rebuilding posts will parse the contained data, but will not re-parse the message itself. 

This was the second suggested change, at the moment this doesn't happen but could easily be changed so that messages get re-parsed thru the rebuild worker.

**Comment by dtdesign** [comment]
Running the full parse process has a significant impact on the performance and for the most part isn't even necessary.

We could think about transforming somewhat okayish looking bbcodes into `<woltlab-metacode>`, but only invoke the bbcode if the attributes match. This way we can handle newly added or updated bbcodes on the fly without the need to rebuild all messages. This however would require a significant change to the inner workings and the assumption made by the entire parsing process, hence this will happen with 3.2 earliest.

**Comment by Morik** [comment]
OK, looking forward to this.
What about the wrapping part, any plans for this ?

**Comment by dtdesign** [comment]
There won't be any change in that direction, using the HTML bbcode is tempting, but leaves too much on the table, in particular messages that contain a mix of HTML and bbcode.

Rebuilding messages will also implicitly revert the metacodes into their basic bbcode representation, resolving the issue of unknown or modified bbcodes at the time of import. This feature is available in WoltLab Suite 5.2 and later.
